ios/SnapyrRnSdk.m

```diff
@@ -68,8 +74,138 @@ @implementation SnapyrRnSdk
+// We provide class methods (i.e. static methods) for end user code to call so they don't need
+// to manage lifecycle. Those class methods use NSNotificationCenter to pass the data along through
+// events; we listen to those events here in the actual instance, where we can then pass data back
+// into React Native.
+-(void)startObserving {
+    // Set up any upstream listeners or background tasks as necessary
+    [[NSNotificationCenter defaultCenter]
+     addObserver:self
+     selector:@selector(handleSnapyrDidRegister:)
+     name:@"snapyrDidRegister"
+     object:nil];
+    
+    [[NSNotificationCenter defaultCenter]
+     addObserver:self
+     selector:@selector(handleSnapyrDidReceiveNotification:)
+     name:@"snapyrDidReceiveNotification"
+     object:nil];
+    
+    [[NSNotificationCenter defaultCenter]
+     addObserver:self
+     selector:@selector(handleSnapyrDidReceiveNotificationResponse:)
+     name:@"snapyrDidReceiveNotificationResponse"
+     object:nil];
+}
+
+// Will be called when this module's last listener is removed, or on dealloc.
+-(void)stopObserving {
+    // Remove upstream listeners, stop unnecessary background tasks
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)handleSnapyrInAppMessage:(SnapyrInAppMessage *)message
+{
+    // Body of events to RN must be JSON serializable so call asDict
+    [self sendEventWithName:@"snapyrInAppMessage" body:[message asDict]];
+}
+
+- (void)handleSnapyrDidRegister:(NSNotification *)notification
+{
+    NSString *token = notification.userInfo[@"token"];
+    [self sendEventWithName:@"snapyrDidRegister" body:token];
+}
+
+- (void)handleSnapyrDidReceiveNotification:(NSNotification *)notification
+{
+    NSDictionary *notif = notification.userInfo[@"notification"];
+    [self sendEventWithName:@"snapyrDidReceiveNotification" body:notif];
+}
+
+- (void)handleSnapyrDidReceiveNotificationResponse:(NSNotification *)notification
+{
+    [self sendEventWithName:@"snapyrDidReceiveNotificationResponse" body:notification.userInfo];
+}
+
+
+- (void)handleTest:(NSNotification *)notification
+{
+    NSString *notif = notification.userInfo[@"notification"];
+    [self sendEventWithName:@"snapyrTest" body:notif];
+}
+
+// Class method the consumer can call when application(_:didRegisterForRemoteNotificationsWithDeviceToken:)
+// has been triggered on the AppDelegate.
+// Decodes token and passes it through NSNotificationCenter so it can be processed by the instance method
+// `handleSnapyrDidRegister`
++ (void)didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
+{
+    NSUInteger dataLength = deviceToken.length;
+    if (dataLength == 0) {
+        return;
+    }
+    
+    const unsigned char *dataBuffer = (const unsigned char *)deviceToken.bytes;
+    NSMutableString *hexString = [NSMutableString stringWithCapacity:(dataLength * 2)];
+    for (int i = 0; i < dataLength; ++i) {
+        [hexString appendFormat:@"%02x", dataBuffer[i]];
+    }
```

Comment:
Wha?

Reply:
Thanks for the review! This is copied straight from our iOS docs (https://developers.snapyr.com/ios-objc.html) - iOS push tokens are received as a binary data format and this is the magic incantation to convert them to a string correctly.

Reply:
Got it, thanks @paulwsmith